File: solutions/haskell/037-group-anagrams.hs
{-
  Problem 37: Group Anagrams (LeetCode 49)
  Difficulty: Med
  Language: Haskell
-}
import Data.List (sort, groupBy, sortBy)
import Data.Ord (comparing)

groupAnagrams :: [String] -> [[String]]
groupAnagrams strs =
  map (map snd) .
  groupBy (\a b -> fst a == fst b) .
  sortBy (comparing fst) $
  map (\s -> (sort s, s)) strs

File: solutions/rust/037-group-anagrams.rs
/**
 * Problem 37: Group Anagrams (LeetCode 49)
 * Difficulty: Med
 * Language: Rust
 */
use std::collections::HashMap;

pub fn group_anagrams(strs: Vec<String>) -> Vec<Vec<String>> {
    strs.into_iter().fold(HashMap::new(), |mut map, s| {
        let mut key: Vec<char> = s.chars().collect();
        key.sort();
        map.entry(key).or_insert_with(Vec::new).push(s);
        map
    }).into_values().collect()
}
